qa/rpc-tests/test_framework/util.py

```diff
@@ -247,17 +249,72 @@ def hex_str_to_bytes(hex_str):
 def str_to_b64str(string):
     return b64encode(string.encode('utf-8')).decode('ascii')
 
+
+connectionTracking = False
+
+def setup_connection_tracking(rpc_connections):
+    global connectionTracking
+    for c in rpc_connections:
+        c.set("net.subversionOverride=%d" % c.auth_service_proxy_instance.nodeNumber)
+    connectionTracking = True
+
+# credit: https://www.python-course.eu/graphs_python.php
+def is_connected(gdict, vertices_encountered = None, start_vertex=None):
+        """ determines if the graph is connected """
+        if vertices_encountered is None:
+            vertices_encountered = set()
+        vertices = list(gdict.keys()) # "list" necessary in Python 3 
+        if not start_vertex:
+            # chosse a vertex from graph as a starting point
+            start_vertex = vertices[0]
+        vertices_encountered.add(start_vertex)
+        if len(vertices_encountered) != len(vertices):
+            for vertex in gdict[start_vertex]:
+                if vertex not in vertices_encountered:
+                    if is_connected(gdict, vertices_encountered, vertex):
+                        return True
+        else:
+            return True
+        return False
+
 def sync_blocks(rpc_connections, wait=1,verbose=1):
     """
     Wait until everybody has the same block count
     """
+    iter=-1
     while True:
         counts = [ x.getblockcount() for x in rpc_connections ]
         if verbose:
             logging.info("sync blocks: " + str(counts))
         if counts == [ counts[0] ]*len(counts):
             break
         time.sleep(wait)
+        iter+=1
+        if connectionTracking and iter&7==0:  # do some other checks to ensure that block sync is possible
+            cxnCount = 0
+            cnxns = [(x.url, x.get("net.subversionOverride")["net.subversionOverride"], x.getpeerinfo()) for x in rpc_connections]
+            graph = { }
+            cnxnCount = 0
+            interestingNodes=[]
+            for (url, node_idx, peers) in cnxns:  # We only care to check the nodes in our rpc_connections
+                # print("%s %s %s" % (url, node_idx, peers))
+                interestingNodes.append(node_idx)
+            # print(interestingNodes)
+
+            for (url, nodeIdx, peers) in cnxns:
+                # port = int(url.split(":")[-1])
+                connectedTo = []
+                for p in peers:
+                    peerIdx = p["subver"]
+                    # print (peerIdx)
+                    if peerIdx in interestingNodes:
+                        connectedTo.append(peerIdx)
+                graph[nodeIdx] = connectedTo
+            print("Node connection graph: ", graph)
+            if not is_connected(graph):
+                print('sync_blocks: bitcoind nodes cannot sync because they are not all connected')
+                pdb.set_trace()
+                raise Exception('sync_blocks: bitcoind nodes cannot sync because they are not all connected')
```

Comment:
do you really need this pdb.set_trace()...I think this is just going to cause a travis hang and timeout if it gets triggered and we'll never get to the exception?

Reply:
thanks, I didn't pass back thru and clean this PR up.